Fixes #12706 - Export ArrayByteBufferPool statistics via JMX.

sbordet opened the pull request:
* Exposed bucket statistics as a list of maps.
* Exposed non-bucket statistic map.
* Fixed get[Direct|Heap]Memory() and exposed getAvailable[Direct|Heap]Memory().

sbordet committed:
Fixes #12706 - Export ArrayByteBufferPool statistics via JMX.

* Exposed bucket statistics as a list of maps.
* Exposed non-bucket statistic map.
* Fixed get[Direct|Heap]Memory() and exposed getAvailable[Direct|Heap]Memory().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -576,6 +627,17 @@ private int evict()
+        private Statistics getStatistics()
+        {
+            List<Pool.Entry<RetainableByteBuffer>> entries = getPool().stream().toList();
+            int inUse = (int)entries.stream().filter(Pool.Entry::isInUse).count();
```
Converting to a list is needlessly wasteful, you could directly stream the pool here with:
```java
int inUse = (int)getPool().stream().filter(Pool.Entry::isInUse).count();
```

sbordet replied:
No, I need both the in-use count and the total count, and I would like them to be consistent, that's why I copy the pool to a list.

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -576,6 +627,17 @@ private int evict()
+            long pooled = _pooled.longValue();
+            long acquires = _acquires.longValue();
+            float hitRatio = acquires == 0 ? Float.NaN : pooled * 100F / acquires;
+            return new Statistics(getCapacity(), inUse, entries.size(), pooled, acquires, _releases.longValue(),
```
Without the list, you can replace `entries.size()` with `getPool().size()` which is as cheap.

sbordet replied:
But not consistent.

lorban replied:
None of those numbers are strictly consistent with each other at any point in time. Having an extra pair of slightly inconsistent numbers won't change anything in practice. _Embrace the race._

lorban reviewed (changes requested):
There's a minor perf issue in `toString`, otherwise LGTM.

wendigo commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ArrayByteBufferPool.java:
```diff
@@ -418,34 +422,81 @@ private long getAvailableByteBufferCount(boolean direct)
+    private List<Map<String, Object>> getBucketsStatistics(boolean direct)
+    {
+        RetainedBucket[] buckets = direct ? _direct : _indirect;
+        return Arrays.stream(buckets).map(b -> b.getStatistics().toMap()).toList();
```
can we filter out buckets that don't have allocations at all?

lorban replied:
That looks like something better left to the caller of this method as we do want to report the buckets that did no allocation.

wendigo commented:
btw. how can I export statistics for two distinct ArrayByteBufferPools - one used in the client and one in the server? It seems that I'm always getting a single pool exported.

sbordet committed:
Updated JMX context attribute for Server in ServerMBean.

This allow to distinguish ByteBufferPools, Schedulers, etc. among multiple servers and from HttpClient or other components.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
> btw. how can I export statistics for two distinct ArrayByteBufferPools - one used in the client and one in the server? It seems that I'm always getting a single pool exported.

The pools should have different `ObjectName`s, so there should be two.

Use `MBeanServer.queryNames(new ObjectName("org.eclipse.jetty.io:type=arraybytebufferpool,*"), null)` to get the two `ObjectNames`.

sbordet committed:
Updates from review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

wendigo commented:
@sbordet how these object pools are getting name assigned? I can't find that. I'm always seeing a single buffer pool in the mbean browser:

<img width="413" alt="Screenshot 2025-01-14 at 17 42 37" src="https://github.com/user-attachments/assets/a86e291c-c155-45d3-b05f-b2ba02a6d673" />

sbordet commented:
It is possible that you are not registering the `HttpClient` instances in JMX?

See: https://jetty.org/docs/jetty/12/programming-guide/arch/jmx.html#enabling-jmx-support
